### PR TITLE
derive_seed request

### PR DESCRIPTION
**Pull request by maackle**

### CODE REVIEW

File: crates/lair_keystore_api/src/lair_client.rs [hunk 1]
```diff
@@ -311,18 +306,40 @@ impl LairClient {
...
-        _dst_deep_lock_passphrase: Option<sodoken::BufRead>,
-        _derivation: Box<[u32]>,
-    ) -> impl Future<Output = LairResult<SeedInfo>> + 'static + Send {
-        async move { unimplemented!() }
+        src_tag: Arc<str>,
+        src_deep_lock_passphrase: Option<BufRead>,
+        dst_tag: Arc<str>,
+        dst_deep_lock_passphrase: Option<BufRead>,
+        derivation_path: Box<[u32]>,
+        // ) -> impl Future<Output = LairResult<SeedInfo>> + 'static + Send {
```
Comment (maackle) [inline crates/lair_keystore_api/src/lair_client.rs:320]: sorry I'll change this back :innocent:

File: crates/lair_keystore_api/src/lair_server/priv_api.rs [hunk 2]
```diff
@@ -233,6 +233,127 @@ pub(crate) fn priv_req_new_seed<'a>(
...
+        }
+
+        let dst_seed = parent;
+
+        // TODO: do we want to take the limits from the source passphrase if possible?
+        // TODO: do we even want to allow a destination passphrase if there is no source passphrase?
+        let (ops_limit, mem_limit) = if let Some(dlp) = src_dlp {
+            (dlp.ops_limit, dlp.mem_limit)
+        } else {
+            let limits = PwHashLimits::current();
```
Comment (neonphog) [inline crates/lair_keystore_api/src/lair_server/priv_api.rs:303]: This call must move up out of the `async move`. If there are any awaits above it, there's a chance the original limits will no longer be in context.

File: crates/lair_keystore_api/src/lair_client.rs [hunk 3]
```diff
@@ -311,18 +306,42 @@ impl LairClient {
     // uhhhh... clippy?? [u32] by itself is not sized... so, yes
     // this *does* have to be Boxed...
     #[allow(clippy::boxed_local)]
+    #[warn(warnings)]
```
Comment (neonphog) [inline crates/lair_keystore_api/src/lair_client.rs:309]:
```suggestion
```

File: crates/lair_keystore_api/src/lair_server/priv_api.rs (hunk 2 above)
Comment (neonphog) [reply crates/lair_keystore_api/src/lair_server/priv_api.rs:303]: aw, crud. it's even worse than that. I think you need to move it outside the `async move` in the main public function, and then pass it into here.

File: crates/lair_keystore_api/src/lair_api/derive_seed.rs [hunk 4]
```diff
@@ -0,0 +1,103 @@
...
+    /// Tag under which to store derived seed.
+    pub dst_tag: Arc<str>,
+
+    /// Optional passphrase with which to deep-lock derived seed.
+    #[serde(skip_serializing_if = "Option::is_none", default)]
+    pub dst_deep_lock_passphrase: Option<DeepLockPassphraseBytes>,
+
+    /// The hierarchy of nested derivations.
+    /// For instance, [0, 2, 1] would specify that the destination seed is the
+    /// 2nd derivation of the 3rd derivation of the 1st derivation of the source seed.
```
Comment (neonphog) [inline crates/lair_keystore_api/src/lair_api/derive_seed.rs:26]:
```suggestion
    /// For instance, [99, 1, 42] would specify that the destination seed is the
    /// 42nd derivation of the 1st derivation of the 99th derivation of the source seed.
```

File: crates/lair_keystore_api/src/lair_api/derive_seed.rs [hunk 5]
```diff
@@ -0,0 +1,103 @@
...
+}
+
+/// Response to a GetEntry request.
+#[derive(Debug, serde::Serialize, serde::Deserialize)]
+#[serde(rename_all = "camelCase")]
+pub struct LairApiResDeriveSeed {
+    /// Msg id to relate request / response.
+    pub msg_id: Arc<str>,
+
+    /// The seed info associated with this seed.
```
Comment (neonphog) [inline crates/lair_keystore_api/src/lair_api/derive_seed.rs:75]:
```suggestion
    /// The seed info associated with the newly derived seed.
```

File: crates/lair_keystore_api/src/lair_api/derive_seed.rs (hunk 4 above)
Comment (neonphog) [reply crates/lair_keystore_api/src/lair_api/derive_seed.rs:26]: oh, i see, you were trying to indicate the zero-based-ness of it. I just thought you got the numbers wrong : )

Well, I'll leave my comment up. I'm not sure we need to be pedantic about the ordinality. Feel free to take or leave this.

**Review by neonphog — CHANGES_REQUESTED** [review]
Couple change requests, but looking good :+1:

File: crates/lair_keystore_api/src/lair_server/priv_api.rs (hunk 2 above)
Comment (neonphog) [reply crates/lair_keystore_api/src/lair_server/priv_api.rs]: Yeah, use new_seed as a template. The [limits are captured at the top](https://github.com/holochain/lair/blob/a4fa49a2b8dcebe5e784ed44d2c22d6044a8abdc/crates/lair_keystore_api/src/lair_client.rs#L189) and then [forwarded to the server](https://github.com/holochain/lair/blob/a4fa49a2b8dcebe5e784ed44d2c22d6044a8abdc/crates/lair_keystore_api/src/lair_client.rs#L206-L207) in the request.

File: crates/lair_keystore_api/src/lair_server/priv_api.rs [hunk 6]
```diff
@@ -233,6 +233,106 @@ pub(crate) fn priv_req_new_seed<'a>(
...
+        };
+
+        let mut parent = src_seed;
+
+        for index in req.derivation_path.iter() {
+            let derived = BufWriteSized::new_mem_locked()?;
+            sodoken::kdf::derive_from_key(
+                derived.clone(),
+                *index as u64,
+                *b"IdnSecKy",
```
Comment (neonphog) [inline crates/lair_keystore_api/src/lair_server/priv_api.rs:285]: Whoops, sorry I missed this earlier, this has to match the seed bundle spec.

```suggestion
                *b"SeedBndl",
```

File: crates/lair_keystore_api/src/lair_server/priv_api.rs [hunk 7]
```diff
@@ -233,6 +233,106 @@ pub(crate) fn priv_req_new_seed<'a>(
...
+                    )
+                    .await?
+            }
+            // create a new seed
+            None => {
+                store
+                    .insert_seed(
+                        dst_seed,
+                        req.dst_tag.clone(),
+                        src_seed_info.exportable,
```
Comment (neonphog) [inline crates/lair_keystore_api/src/lair_server/priv_api.rs:317]: ... someday we'll probably want to be able to change the exportable and limits when deriving... but out of scope until we have an actual need for it : )

**Review by neonphog — CHANGES_REQUESTED** [review]
Sorry I missed the context bytes needing to match. After that, this looks good to go.

**Review by neonphog — APPROVED** [review]

**Pull request merged**